pkg/bitrot/bitrot.go

```diff
@@ -0,0 +1,151 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package bitrot provides algorithms for detecting data corruption.
+package bitrot
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+)
+
+// Mode specifies how the Hash will be used - either for protection or for verification.
+type Mode bool
+
+const (
+	// Protect indicates that the Hash is used for computing a checksum.
+	Protect Mode = true
+	// Verify indicates that the Hash is used for verifying a previously computed checksum.
+	Verify Mode = !Protect
+)
+
+// Algorithm identifies a bitrot detection algorithm that is implemented in another
+// package.
+type Algorithm uint
+
+const (
+	// Poly1305 specifies the poly1305 polynomial authenticator
+	Poly1305 Algorithm = iota
+
+	// GHASH specifies the GHASH (AES-GCM) polynomial authenticator
+	GHASH
+
+	// BLAKE2b512 specifies the BLAKE2b-512 hash function
+	BLAKE2b512
+
+	// SHA256 specifies the SHA256 hash function
+	SHA256
+
+	// add algorithms here
+
+	// UnknownAlgorithm indicates an unknown/invalid algorithm
+	UnknownAlgorithm
+)
+
+var keysizes = []int{
+	Poly1305:   32,
+	GHASH:      32,
+	BLAKE2b512: 0,
+	SHA256:     0,
+}
+
+var names = []string{
+	Poly1305:   "poly1305",
+	GHASH:      "ghash",
+	BLAKE2b512: "blake2b",
```

Comment:
Consider renaming constant `BLAKE2b512' to 'BLAKE2b' in order to match the string name (it is now the odd one out here, and blake2b is what is current written in the json).